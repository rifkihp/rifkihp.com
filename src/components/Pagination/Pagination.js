import classNames from 'classnames/bind';
import { Link } from 'gatsby';
import React from 'react';

import styles from './Pagination.module.scss';

const cx = classNames.bind(styles);

const Pagination = ({ classes, currentPage, pagePath, hasNextPage, hasPrevPage, numPages }) => {
  if (!hasNextPage && !hasPrevPage) {
    return null;
  }

  const prevClassName = cx({
    'pagination__prev-link': true,
    'pagination__prev-link--disable': !hasPrevPage,
  });

  const nextClassName = cx({
    'pagination__next-link': true,
    'pagination__next-link--disable': !hasNextPage,
  });

  let pageNumbers = [];

  // If there are <= 7 pages, we'll never need an ellipsis.
  if (numPages <= 7) {
    for (let i = 1; i <= numPages; i++) {
      pageNumbers.push(i);
    }
  } else if (currentPage <= 4) {
    // 1 2 3 4 5 ... n
    pageNumbers = [1, 2, 3, 4, 5, '…', numPages];
  } else if (currentPage < numPages - 3) {
    // 1 ... c-1 c c+1 ... n
    pageNumbers = [1, '…', currentPage - 1, currentPage, currentPage + 1, '…', numPages];
  } else {
    // 1 ... n-4 n-3 n-2 n-1 n
    pageNumbers = [1, '…', numPages - 4, numPages - 3, numPages - 2, numPages - 1, numPages];
  }

  return (
    <div className={`${styles['pagination']} ${classes || ''}`}>
      <div className={styles['pagination__prev']}>
        <Link
          rel="prev"
          to={hasPrevPage ? pagePath(currentPage - 1) : '/#'}
          className={prevClassName}
        >
          ← PREV
        </Link>
      </div>
      <ul className={styles['pagination__list-container']}>
        {pageNumbers.map((pn, i) => (
          <li key={`pagenum-${i}`}>
            {pn === '…' ? (
              <p>…</p>
            ) : (
              <Link to={pagePath(pn)} className={pn === currentPage ? styles['selected'] : ''}>
                {pn}
              </Link>
            )}
          </li>
        ))}
      </ul>
      <div className={styles['pagination__next']}>
        <Link
          rel="next"
          to={hasNextPage ? pagePath(currentPage + 1) : '/#'}
          className={nextClassName}
        >
          NEXT →
        </Link>
      </div>
    </div>
  );
};

export default Pagination;
