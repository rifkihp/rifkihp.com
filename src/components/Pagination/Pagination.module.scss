@import '../../assets/scss/variables';
@import '../../assets/scss/mixins';

.pagination {
  @include margin-top(2);
  display: flex;
  justify-content: center;

  &__prev {
    flex-grow: 1;
    text-align: left;

    &-link {
      color: $color-primary;
      font-size: 26px;
      font-weight: bold;

      &:hover,
      &:focus {
        color: $color-secondary;
      }

      &--disable {
        pointer-events: none;
        color: $color-gray-shadow;
      }
    }
  }

  &__next {
    flex-grow: 1;
    text-align: right;

    &-link {
      color: $color-primary;
      font-size: 26px;
      font-weight: bold;

      &:hover,
      &:focus {
        color: $color-secondary;
      }

      &--disable {
        pointer-events: none;
        color: $color-gray-shadow;
      }
    }
  }

  &__prev, &__next {
    display: none;
  }

  &__list-container {
    @include margin(0, 0);

    li {
      display: inline;
      @include margin(0, 0.15);

      a {
        font-weight: 600;
        display: inline;
        @include padding(0.25, 0.25);
      }

      a.selected {
        color: var(--color-white);
        background-color: $color-primary;
        @include padding(0.25, 0.4);
        border-radius: 3px;
      }

      p {
        display: inline;
        @include padding(0.25, 0);
      }
    }
  }
}

@include breakpoint-md {
  .pagination {
    &__prev, &__next {
      display: unset;
    }
  }
}
