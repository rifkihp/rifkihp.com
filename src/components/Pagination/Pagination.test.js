import React from 'react';
import renderer from 'react-test-renderer';

import Pagination from './Pagination';

describe('Pagination', () => {
  const props = {
    pagePath: p => `/test/${p}`,
    hasNextPage: true,
    hasPrevPage: true,
  };

  it('renders correctly on page 2 / 5', () => {
    const tree = renderer.create(<Pagination {...props} currentPage={2} numPages={5} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly on page 6 / 10', () => {
    const tree = renderer.create(<Pagination {...props} currentPage={6} numPages={10} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
