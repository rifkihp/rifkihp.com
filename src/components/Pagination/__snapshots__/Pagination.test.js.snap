// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders correctly on page 2 / 5 1`] = `
<div
  className="pagination "
>
  <div
    className="pagination__prev"
  >
    <Link
      className="pagination__prev-link"
      rel="prev"
      to="/test/1"
    >
      ← PREV
    </Link>
  </div>
  <ul
    className="pagination__list-container"
  >
    <li>
      <Link
        className=""
        to="/test/1"
      >
        1
      </Link>
    </li>
    <li>
      <Link
        className="selected"
        to="/test/2"
      >
        2
      </Link>
    </li>
    <li>
      <Link
        className=""
        to="/test/3"
      >
        3
      </Link>
    </li>
    <li>
      <Link
        className=""
        to="/test/4"
      >
        4
      </Link>
    </li>
    <li>
      <Link
        className=""
        to="/test/5"
      >
        5
      </Link>
    </li>
  </ul>
  <div
    className="pagination__next"
  >
    <Link
      className="pagination__next-link"
      rel="next"
      to="/test/3"
    >
      NEXT →
    </Link>
  </div>
</div>
`;

exports[`Pagination renders correctly on page 6 / 10 1`] = `
<div
  className="pagination "
>
  <div
    className="pagination__prev"
  >
    <Link
      className="pagination__prev-link"
      rel="prev"
      to="/test/5"
    >
      ← PREV
    </Link>
  </div>
  <ul
    className="pagination__list-container"
  >
    <li>
      <Link
        className=""
        to="/test/1"
      >
        1
      </Link>
    </li>
    <li>
      <p>
        …
      </p>
    </li>
    <li>
      <Link
        className=""
        to="/test/5"
      >
        5
      </Link>
    </li>
    <li>
      <Link
        className="selected"
        to="/test/6"
      >
        6
      </Link>
    </li>
    <li>
      <Link
        className=""
        to="/test/7"
      >
        7
      </Link>
    </li>
    <li>
      <p>
        …
      </p>
    </li>
    <li>
      <Link
        className=""
        to="/test/10"
      >
        10
      </Link>
    </li>
  </ul>
  <div
    className="pagination__next"
  >
    <Link
      className="pagination__next-link"
      rel="next"
      to="/test/7"
    >
      NEXT →
    </Link>
  </div>
</div>
`;
